### PR TITLE
JAMES-3367 Add capability mandatory checks

### PR DESCRIPTION
**Pull request by LanKhuat**

### CODE REVIEW

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EchoMethodContract.scala [hunk 1]
```diff
@@ -128,4 +128,31 @@ trait EchoMethodContract {
 
     assertThatJson(response).isEqualTo(RESPONSE_OBJECT_WITH_UNSUPPORTED_METHOD)
   }
+
+  @Test
+  def echoMethodShouldReturnInvalidArgumentsWhenMissingCoreCapability(): Unit = {
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EchoMethodContract.scala:133]:
```suggestion
  def echoMethodShouldReturnUnknownMethodWhenMissingCoreCapability(): Unit = {
```

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxGetMethodContract.scala [hunk 2]
```diff
@@ -1406,4 +1406,80 @@ trait MailboxGetMethodContract {
       .body(s"$SECOND_MAILBOX.name", equalTo(DefaultMailboxes.TRASH))
       .body(s"$SECOND_MAILBOX.sortOrder", equalTo(60))
   }
+
+  @Test
+  def getShouldReturnInvalidArgumentsWhenMissingOneCapability(): Unit = {
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxGetMethodContract.scala:1411]: Same remark as above for all tests: `InvalidArguments` -> `UnknownMethod`

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/JMAPApiRoutes.scala [hunk 3]
```diff
@@ -164,10 +164,22 @@ class JMAPApiRoutes (val authenticator: Authenticator,
...
     SMono.justOrEmpty(methodsByName.get(invocation.methodName))
-      .flatMap(method => SMono.fromPublisher(method.process(capabilities, invocation, mailboxSession, processingContext)))
+      .flatMap(method => validate(capabilities, method.requiredCapabilities)
+      .fold(e => SMono.just((Invocation.error(ErrorCode.UnknownMethod, e.description, invocation.methodCallId), processingContext)),
+        _ => SMono.fromPublisher(method.process(capabilities, invocation, mailboxSession, processingContext))
+      ))
       .onErrorResume(throwable => SMono.just((Invocation.error(ErrorCode.ServerFail, throwable.getMessage, invocation.methodCallId), processingContext)))
       .switchIfEmpty(SMono.just((Invocation.error(ErrorCode.UnknownMethod, invocation.methodCallId), processingContext)))
 
+  private def validate(capabilities: Set[CapabilityIdentifier], requiredCapabilities: Capabilities): Either[MissingCapabilityException, Unit] = {
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/JMAPApiRoutes.scala:174]:
```suggestion
  private def validateCapabilities(capabilities: Set[CapabilityIdentifier], requiredCapabilities: Capabilities): Either[MissingCapabilityException, Unit] = {
```

**Review by Arsnael — APPROVED** [review]

**Review by chibenwa — APPROVED** [review]

**Comment by chibenwa** [comment]
```
[8c01ed25bc5780d78f409e02c73ede17b4e5a4c3] RFC-8621 with random error when processing request 
[8c01ed25bc5780d78f409e02c73ede17b4e5a4c3] - should return 200, with serverFail error, others method call proceed normally *** FAILED ***
[8c01ed25bc5780d78f409e02c73ede17b4e5a4c3]   net.javacrumbs.jsonunit.core.internal.JsonAssertError: JSON documents are different:
[8c01ed25bc5780d78f409e02c73ede17b4e5a4c3] Different value found in node "methodResponses[0][1].description", expected: <"Unexpected Exception occur, the others method may proceed normally"> but was: <null>.
[8c01ed25bc5780d78f409e02c73ede17b4e5a4c3] Different value found in node "methodResponses[1][0]", expected: <"Core/echo"> but was: <"error">.
[8c01ed25bc5780d78f409e02c73ede17b4e5a4c3] Different keys found in node "methodResponses[1][1]", missing: "methodResponses[1][1].arg1","methodResponses[1][1].arg2", extra: "methodResponses[1][1].description","methodResponses[1][1].type", expected: <{"arg1":"arg1data","arg2":"arg2data"}> but was: <{"description":null,"type":"serverFail"}>
[8c01ed25bc5780d78f409e02c73ede17b4e5a4c3]   at net.javacrumbs.jsonunit.core.internal.ExceptionUtils.createException(ExceptionUtils.java:50)
[8c01ed25bc5780d78f409e02c73ede17b4e5a4c3]   at net.javacrumbs.jsonunit.core.internal.Diff.failIfDifferent(Diff.java:613)
[8c01ed25bc5780d78f409e02c73ede17b4e5a4c3]   at net.javacrumbs.jsonunit.assertj.JsonAssert.isEqualTo(JsonAssert.java:106)
[8c01ed25bc5780d78f409e02c73ede17b4e5a4c3]   at org.apache.james.jmap.routes.JMAPApiRoutesTest.$anonfun$new$12(JMAPApiRoutesTest.scala:480)
[8c01ed25bc5780d78f409e02c73ede17b4e5a4c3]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[8c01ed25bc5780d78f409e02c73ede17b4e5a4c3]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[8c01ed25bc5780d78f409e02c73ede17b4e5a4c3]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[8c01ed25bc5780d78f409e02c73ede17b4e5a4c3]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[8c01ed25bc5780d78f409e02c73ede17b4e5a4c3]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[8c01ed25bc5780d78f409e02c73ede17b4e5a4c3]   at org.scalatest.flatspec.AnyFlatSpecLike$$anon$5.apply(AnyFlatSpecLike.scala:1683)
```

Looks like there's a unit test to fix...

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Comment by Arsnael** [comment]
Merged

**Pull request closed without merging**